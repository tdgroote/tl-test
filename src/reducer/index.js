import {
    SUCCESS_PLACE,
    SUCCESS_REMOVE
} from '../constants';

import {
    combineObjs,
    argToObjKey,
    removeById,
    concat
} from '../util';

import { noop, flowRight as compose, identity, flip } from 'lodash';

const ordersActionHandler = handler => compose(
    argToObjKey('orders'),
    (payload, { orders }) => handler(payload, orders)
);

const initialState = {
    products: [],
    customers: [],
    orders: [],
    loadingInitialData: true,
    alert: ''
};

const alert = alert => newState => combineObjs(newState, { alert });
const removeOrder = ordersActionHandler(flip(removeById));

const actionHandlers = {
    initialLoadingDone: () => ({ loadingInitialData: false }),
    setProducts: argToObjKey('products'),
    setCustomers: argToObjKey('customers'),
    alert: argToObjKey('alert'),
    closeAlert: alert(''),
    setOrders: ordersActionHandler(identity),
    createOrder: ordersActionHandler(flip(concat)),
    placeOrder: compose(alert(SUCCESS_PLACE), removeOrder),
    removeOrder: compose(alert(SUCCESS_REMOVE), removeOrder),
    updateOrder: ordersActionHandler((changedOrder, orders) =>
        orders.map(order => order.id === changedOrder.id ? changedOrder : order)
    )
};

export default (state = initialState, action) => combineObjs(
    state,
    (actionHandlers[action.type] || noop)(action.payload, state)
);
