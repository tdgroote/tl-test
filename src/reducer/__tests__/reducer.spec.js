import reducer from '../';

const action = (type, payload) => ({ type, payload });

describe('reducer', () => {
    it('initializes the state', () => {
        expect(reducer(undefined, {})).toMatchSnapshot();
    });

    it('updates the loading state', () => {
        const state = { loadingInitialData: true };
        const expected = false;
        const result = reducer(state, action('initialLoadingDone'));

        expect(result.loadingInitialData).toEqual(expected);
    });

    it('sets products', () => {
        const state = {
            products: [1, 2, 3]
        };
        const expected = [3, 4, 5];
        const result = reducer(state, action('setProducts', [3, 4, 5]));

        expect(result.products).toEqual(expected);
    });

    it('sets customers', () => {
        const state = {
            customers: [1, 2, 3]
        };
        const expected = [3, 4, 5];
        const result = reducer(state, action('setCustomers', [3, 4, 5]));

        expect(result.customers).toEqual(expected);
    });

    it('sets orders', () => {
        const state = {
            orders: [1, 2, 3]
        };
        const expected = [3, 4, 5];
        const result = reducer(state, action('setOrders', [3, 4, 5]));

        expect(result.orders).toEqual(expected);
    });

    it('removes orders', () => {
        const state = {
            orders: [{ id: 1 }, { id: 2 }, { id: 3 }]
        };
        const expected = [{ id: 1 }, { id: 3 }];
        const result = reducer(state, action('removeOrder', 2));

        expect(result.orders).toEqual(expected);
    });

    it('places orders', () => {
        const state = {
            orders: [{ id: 1 }, { id: 2 }, { id: 3 }]
        };
        const expected = [{ id: 1 }, { id: 2 }];
        const result = reducer(state, action('placeOrder', 3));

        expect(result.orders).toEqual(expected);
    });

    it('creates new orders', () => {
        const state = {
            orders: [{ id: 1 }, { id: 2 }]
        };
        const expected = [{ id: 1 }, { id: 2 }, { id: 3 }];
        const result = reducer(state, action('createOrder', {id: 3}));

        expect(result.orders).toEqual(expected);
    });

    it('updates orders', () => {
        const state = {
            orders: [{ id: 1 }, { id: 2 }, { id: 3 }]
        };
        const expected = [{ id: 1 }, { id: 2, newField: 'newValue' }, { id: 3 }];
        const result = reducer(state, action('updateOrder', {
            id: 2,
            newField: 'newValue'
        }));

        expect(result.orders).toEqual(expected);
    });

    it('alerts', () => {
        const state = {};
        const expected = 'message';
        const result = reducer(state, action('alert', 'message'));
        expect(result.alert).toEqual(expected);
    });

    it('closes alerts', () => {
        const state = { alert: 'message' };
        const expected = '';
        const result = reducer(state, action('closeAlert'));
        expect(result.alert).toEqual(expected);
    });
});
