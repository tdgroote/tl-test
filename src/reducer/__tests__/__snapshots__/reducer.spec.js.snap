// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer initializes the state 1`] = `
Object {
  "alert": "",
  "customers": Array [],
  "loadingInitialData": true,
  "orders": Array [],
  "products": Array [],
}
`;
